File: api/keywords-finalise.py
import pandas as pd

from api._lib.blob_io import load_df, save_df
from api._lib.core.keyword_builder import make_keywords
from api._lib.handler_utils import BaseHandler, get_processed_df, records


class handler(BaseHandler):
    def do_POST(self):
        try:
            body = self._json_body()
            raw_df = load_df(body["rawDfUrl"])
            params = {**body.get("state", {}), "colourMapBlobUrl": body.get("colourMapBlobUrl")}
            processed = get_processed_df(raw_df, params)

            combined_tables = []
            for i, combo in enumerate(body.get("combos", [])):
                fields = [f for f in combo.get("fields", []) if f in processed.columns]
                table = make_keywords(processed, fields, bool(combo.get("splitAmpersand", False)))
                list_name = combo.get("name") or f"List {i + 1}"
                if not table.empty:
                    table.insert(0, "list_name", list_name)
                    combined_tables.append(table)

            combined = pd.concat(combined_tables, ignore_index=True) if combined_tables else pd.DataFrame()
            combined_url = save_df(combined, "combined_keywords")

            self._send_json(200, {
                "combinedDfBlobUrl": combined_url,
                "preview": records(combined),
                "totalKeywords": len(combined),
            })
        except Exception as e:
            self._send_json(500, {"error": str(e)})

File: api/keywords-finalise.ts
import type { VercelRequest, VercelResponse } from '@vercel/node'
import { sendError, records } from '../api/_lib/handlerUtils.js'
import { loadRows, saveRows } from '../api/_lib/sessionStore.js'
import { getProcessedRows } from '../api/_lib/pipeline.js'
import { makeKeywords } from '../api/_lib/core/keywordBuilder.js'
import type { Combo } from '../api/_lib/presets.js'
import type { Row } from '../api/_lib/core/feedProcessor.js'

export default async function handler(req: VercelRequest, res: VercelResponse) {
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' })

  try {
    const { rawDfUrl, session: sessionId, state = {}, combos = [] } = req.body as {
      rawDfUrl: string
      session: string
      state?: Record<string, unknown>
      combos: Combo[]
    }

    const raw = await loadRows(rawDfUrl)
    const processed = await getProcessedRows(raw, state as never)
    const allCols = Object.keys(processed[0] ?? {})

    const allRows: Row[] = []
    for (const combo of combos) {
      const fields = (combo.fields ?? []).filter(f => allCols.includes(f))
      const table = makeKeywords(processed, fields, Boolean(combo.splitAmpersand))
      const listName = combo.name || 'List'
      for (const row of table) allRows.push({ ...row, list_name: listName })
    }

    const combinedRef = await saveRows(sessionId, 'combined', allRows)

    res.status(200).json({
      combinedDfBlobUrl: combinedRef,
      preview: records(allRows),
      totalKeywords: allRows.length,
    })
  } catch (err) {
    sendError(res, err)
  }
}
